src/equations/gradient_equations_2d.jl

```diff
@@ -0,0 +1,46 @@
+
+@doc raw"""
+    GradientEquations2D
+
+The gradient equations
+```math
+q^d - \partial_d u = 0
+```
+in direction `d` in two space dimensions as required for, e.g., the Bassi & Rebay 1 (BR1) or the
+local discontinuous Galerkin (LDG) schemes.
+"""
+struct GradientEquations2D{RealT<:Real, NVARS} <: AbstractGradientEquations{2, NVARS}
+  orientation::Int
+end
+
+GradientEquations2D(::Type{RealT}, nvars, orientation) where RealT = GradientEquations2D{RealT, nvars}(orientation)
+GradientEquations2D(nvars, orientation) = GradientEquations2D(Float64, nvars, orientation)
+
+
+get_name(::GradientEquations2D) = "GradientEquations2D"
+varnames(::typeof(cons2cons), equations::GradientEquations2D) = SVector(ntuple(v -> "gradient_"*string(v), nvariables(equations)))
+varnames(::typeof(cons2prim), equations::GradientEquations2D) = varnames(cons2cons, equations)
+
+# Set initial conditions at physical location `x` for time `t`
+"""
+    initial_condition_constant(x, t, equations::GradientEquations2D)
+
+A constant initial condition to test free-stream preservation.
+"""
+function initial_condition_constant(x, t, equations::GradientEquations2D)
+  return SVector(ntuple(v -> zero(eltype(x)), nvariables(equations)))
+end
+
+
+# Pre-defined source terms should be implemented as
+# function source_terms_WHATEVER(u, x, t, equations::GradientEquations2D)
+
+
+# Calculate 1D flux in for a single point
+@inline function calcflux(u, orientation, equations::GradientEquations2D)
```

Comment:
This naming convention seems odd to me. This is not the viscous flux because that lives elsewhere, e.g. in HeatEquation2D. Is this used to select which set of variables with be used to compute the gradients?

For linear advection diffusion or the heat equation the choice is usually the obvious one of `u` but for something like the Navier-Stokes equations we have different sets of variables with which we can compute the gradient, e.g. `(rho,u,v,p)` or `(rho, rho_u, rho_v, E)` or `(u,v,T)` or the entropy variables. Will this `gradient_equations_2d.jl` be extended in this way?

Reply:
> This naming convention seems odd to me. This is not the viscous flux because that lives elsewhere, e.g. in HeatEquation2D. Is this used to select which set of variables with be used to compute the gradients?

No, at the moment this is a kludge since the `GradientEquations2D` are solved in all direction even though by definition the flux is non-zero only in one direction.

> Will this gradient_equations_2d.jl be extended in this way?

Until you asked I had no plans to do so, no. But I think it should not be too hard to do. How is this handled in Fluxo?

Reply:
> Until you asked I had no plans to do so, no. But I think it should not be too hard to do. How is this handled in Fluxo?

Basically there is a flag to indicate which set of variables one uses to compute the gradients. The available options in Fluxo are primitive, conservative or entropy variables. Then  from these gradients we reverse engineer the expressions we need to compute the viscous flux components. For example, we would need u_x which for the primitive variables is just a copy (it is already computed) but for the conservative variables we have rho_x and (rho u)_x computed and thus can obtain u_x = ( (rho u)_x - rho_x u )/rho. It is done in this way such that the viscous flux routine is agnostic (for the most part) as to which form of gradients we use

Reply:
We could think of providing `cons2auxvars` and `auxvars2x`, where `auxvars` are the variables in which we want to compute the gradients and `x` the variables in which we want to use them inside our viscous fluxes. So I think this is something that is generally possible.

Unless such a conversion feature is crucial to have from the start, however, I'd rather leave this out of this PR and implement it a later point once someone really wants to use anything other than gradients in the conservative variables. Would that be OK for you?

Reply:
Yeah, that is okay by me. I just wanted to throw in my outlook on this routine structure because I know that implementing viscous terms is new

Reply:
Yes, thanks a lot for the feedback! I noted your point in the description [above](https://github.com/trixi-framework/Trixi.jl/pull/378#issue-536620073).